### PR TITLE
Update to version 0.4 of LibHac

### PR DESCRIPTION
**Pull request by Thealexbarney**
Use the `IFileSystem`, `IFile`, `IDirectory`, etc. interfaces from LibHac
Use DirectorySaveDataFileSystem for savedata

### CODE REVIEW

File: Ryujinx.HLE/HOS/Services/FspSrv/IDirectory.cs [hunk 1]
```diff
@@ -15,40 +13,26 @@ class IDirectory : IpcService, IDisposable
...
-            {
-                _directoryEntries.AddRange(provider.GetDirectories(directoryPath));
-            }
-
-            if ((flags & 2) != 0)
-            {
-                _directoryEntries.AddRange(provider.GetFiles(directoryPath));
-            }
+            _provider     = directory;
+            Path = directory.FullPath;
```
Comment (emmauss) [inline Ryujinx.HLE/HOS/Services/FspSrv/IDirectory.cs]: align

File: Ryujinx.HLE/HOS/Services/FspSrv/IDirectory.cs [hunk 2]
```diff
@@ -58,41 +42,42 @@ public long Read(ServiceCtx context)
             long bufferLen      = context.Request.ReceiveBuff[0].Size;
 
             int maxReadCount = (int)(bufferLen / DirectoryEntrySize);
+            int readCount = 0;
```
Comment (emmauss) [inline Ryujinx.HLE/HOS/Services/FspSrv/IDirectory.cs]: align

File: Ryujinx.HLE/HOS/Services/FspSrv/IFile.cs [hunk 3]
```diff
@@ -28,8 +27,8 @@ public IFile(Stream baseStream, string hostPath)
                 { 4, GetSize }
             };
 
-            _baseStream = baseStream;
-            HostPath   = hostPath;
+            _baseFile = baseFile;
+            Path = LibHac.Fs.PathTools.Normalize(path);
```
Comment (emmauss) [inline Ryujinx.HLE/HOS/Services/FspSrv/IFile.cs]: align

File: Ryujinx.HLE/HOS/Services/FspSrv/IFileSystem.cs [hunk 4]
```diff
@@ -53,105 +50,171 @@ public long CreateFile(ServiceCtx context)
         {
             string name = ReadUtf8String(context);
 
-            long mode = context.RequestData.ReadInt64();
-            int  size = context.RequestData.ReadInt32();
+            int mode = context.RequestData.ReadInt32();
+            long size = context.RequestData.ReadInt64();
```
Comment (emmauss) [inline Ryujinx.HLE/HOS/Services/FspSrv/IFileSystem.cs]: align

**Comment by emmauss** [comment]
Done with an initial check. just some alignment issues.

File: Ryujinx.HLE/HOS/Services/FspSrv/IFileSystemProxy.cs [hunk 5]
```diff
@@ -224,57 +223,44 @@ public void LoadSaveDataFileSystem(ServiceCtx context)
...
+
+            MakeObject(context, new IFileSystem(saveFileSystem));
         }
 
         private long OpenNsp(ServiceCtx context, string pfsPath)
         {
-            FileStream pfsFile = new FileStream(pfsPath, FileMode.Open, FileAccess.Read);
-            Pfs        nsp     = new Pfs(pfsFile.AsStorage());
+            LocalStorage storage    = new LocalStorage(pfsPath, FileAccess.Read, FileMode.Open);
+            PartitionFileSystem nsp = new PartitionFileSystem(storage);
```
Comment (AcK77) [inline Ryujinx.HLE/HOS/Services/FspSrv/IFileSystemProxy.cs]: Align vars names

**Pull request merged**